src/bin/everest.rs

```diff
@@ -0,0 +1,294 @@
+        // - the timeline got updated
+        select! {
+            result = receiver.recv() => {
+                match result {
+                    None => break, // sender was dropped
+                    Some(new_timeline) => timeline = new_timeline,
+                }
+            }
+            () = wait_for_next_start(&clock, &timeline) => {
+                //  fall through
+            }
+        };
+
+        let now = clock.now();
+
+        // normally, we expect this to return, but due to e.g. time synchronization, we may wake up
+        // without any interval to actually send upstream.
+        let Some(current) = timeline.at_datetime(&now) else {
+            continue;
+        };
+
+        let mut schedule = Vec::new();
+        let mut valid_until = None;
+
+        for (range, interval) in timeline.iter() {
+            // skip anything that is already complete
+            if range.end < now {
+                continue;
+            }
+
+            valid_until = Ord::max(valid_until, Some(range.end));
+
+            let entry = ScheduleResEntry {
+                timestamp: range.start,
+                limits_to_root: LimitsRes::try_from_event_values(interval.value_map).unwrap(),
+            };
+
+            schedule.push(entry);
+        }
+
+        let opt_limits = LimitsRes::try_from_event_values(current.1.value_map);
+        if let (Some(valid_until), Some(limits_root_side)) = (valid_until, opt_limits) {
+            let enforced_limits = EnforcedLimits {
+                uuid: Uuid::new_v4().to_string(),
+                valid_until,
+                limits_root_side,
+                schedule,
+            };
+
+            let Ok(()) = sender.send(enforced_limits).await else {
+                break;
+            };
+        }
+    }
+}
+
+// https://github.com/tdittr/everest-core/blob/openadr/types/energy.yaml#L213
+#[derive(Debug, Clone, PartialEq)]
+struct EnforcedLimits {
+    uuid: String,
+    valid_until: chrono::DateTime<chrono::Utc>,
+    limits_root_side: LimitsRes,
+    schedule: Vec<ScheduleResEntry>,
+}
+
+// https://github.com/tdittr/everest-core/blob/openadr/types/energy.yaml#L125
+#[derive(Debug, Clone, Copy, PartialEq)]
+struct ScheduleResEntry {
+    timestamp: chrono::DateTime<chrono::Utc>,
+    limits_to_root: LimitsRes,
+}
+
+// https://github.com/tdittr/everest-core/blob/openadr/types/energy.yaml#L46
+#[derive(Debug, Clone, Copy, PartialEq)]
+struct LimitsRes {
+    // NOTE: that W is uppercase if we ever need to serialize this!
+    total_power_w: f64,
+}
+
+impl LimitsRes {
+    fn try_from_event_values(values: &[EventValuesMap]) -> Option<Self> {
+        for EventValuesMap { value_type, values } in values {
+            if let EventType::ImportCapacityLimit = value_type {
+                let total_power_w = match &values[..] {
+                    [Value::Integer(value)] => *value as f64,
+                    [Value::Number(value)] => *value,
```

Comment:
I think the integer case is required because of how `Value`s get parsed: the integer case comes first in the definition, which would mean that anything that looks like an integer is parsed as an integer, even if downstream we really want a float.

Reply:
good catch!